components/automate-ui/src/app/modules/user/user-details/user-details.component.ts

```diff
@@ -120,13 +102,107 @@ export class UserDetailsComponent implements OnInit, OnDestroy {
     this.isDestroyed.complete();
   }
 
+  public savePassword(): void {
+    this.userDetails.savePassword();
+  }
+
+  public saveDisplayName(): void {
+    this.userDetails.saveDisplayName();
+  }
+
+  public onSelectedTab(event: { target: { value: UserTabName } }): void {
+    this.tabValue = event.target.value;
+    // Drop the previous fragment and add the incoming fragment.
+    this.router.navigate([this.url.split('#')[0]], { fragment: event.target.value });
+  }
+}
+
+interface UserDetails {
```

Comment:
Below there are three classes that implement this interface that represent the three cases from above.

Reply:
I like the approach! Have you considered abstract classes? Seems like that would help you make it even more tidy. What it gives you "for free" is that the methods that are invariant do not have to be duplicated. You define just those methods in the abstract class, leaving the customized methods to be defined in the derived classes. Reference: https://www.typescriptlang.org/docs/handbook/classes.html#abstract-classes

Reply:
I don't think I have done abstract classes in javascript before. I will give it a try.